Dockerfile

```diff
@@ -2,7 +2,18 @@ FROM debian:stretch-slim
 COPY create_pam_debians.sh /usr/local/bin/
 RUN bash -ex create_pam_debians.sh
 
+FROM debian:stretch-slim
+WORKDIR /usr/local/repos/pam
+COPY --from=0 /tmp/*.deb ./
+COPY create_local_pam_repo.sh /usr/local/bin
+RUN bash -ex create_local_pam_repo.sh
+
 FROM debian:stretch-slim
 COPY --from=0 /tmp/*.deb /
+COPY --from=1 /etc/apt/ /etc/apt/
+COPY --from=1 /usr/local/repos/pam /usr/local/repos/pam
 COPY setup.sh /usr/local/bin/
-RUN bash -ex setup.sh
+RUN bash -ex setup.sh && rm /usr/local/bin/setup.sh
+
+FROM scratch
+COPY --from=2 / /
```

Comment:
I understand that this reduces the size of the image - as there is only one layer now. However, it also removes any association to the `debian:stretch-slim` base image - which could be a shared layer with other images on a system. So, now its a totally unique layer to the pihole, potentially increasing the total space usage of docker for minimal layer size gains. Would this not be a very similar image size, while still keeping cache benifits of shared layers?

```
FROM debian:stretch-slim
COPY --from=2 / /
```

Reply:
```
FROM debian:stretch-slim
COPY --from=2 / /
```
Would double image size. The `COPY` directive just performs a dumb copy, it doesn't check if it's coping files that are exact duplicates to those in the previous layer.

There are two options to keep the `debian:stretch-slim` association:

- A) drop off that last build stage and use artifact from stage 2.
- B) figure out what gets touched by the new pkg installs and apt-hold and manually specify that only those files get copied.

Option A introduces only a minor size regression of 2MB uncompressed as is (a few tweaks to stage 2 can ameliorate this) and would be my recommendation if you wanted to open a Issue/PR for this. Option B is much more headache than benefit.

Reply:
Thanks for the explanation @DistractionRectangle